sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java

```diff
@@ -143,75 +147,75 @@ public void initialize(
   /**
    * Initialize columnar batch by setting required schema and partition information.
    * With this information, this creates ColumnarBatch with the full schema.
+   *
+   * @param orcSchema Schema from ORC file reader.
+   * @param requiredFields All the fields that are required to return, including partition fields.
+   * @param requestedDataColIds Requested column ids from orcSchema. -1 if not existed.
+   * @param requestedPartitionColIds Requested column ids from partition schema. -1 if not existed.
+   * @param partitionValues Values of partition columns.
    */
   public void initBatch(
       TypeDescription orcSchema,
-      int[] requestedColIds,
       StructField[] requiredFields,
-      StructType partitionSchema,
+      int[] requestedDataColIds,
+      int[] requestedPartitionColIds,
       InternalRow partitionValues) {
     batch = orcSchema.createRowBatch(capacity);
     assert(!batch.selectedInUse); // `selectedInUse` should be initialized with `false`.
-
+    assert(requiredFields.length == requestedDataColIds.length);
+    assert(requiredFields.length == requestedPartitionColIds.length);
+    // If a required column is also partition column, use partition value and don't read from file.
+    for (int i = 0; i < requiredFields.length; i++) {
+      if (requestedPartitionColIds[i] != -1) {
+        requestedDataColIds[i] = -1;
+      }
+    }
```

Comment:
Does this loop work as expected? The intention seems to be clear, but [here](https://github.com/apache/spark/pull/23387/files#diff-8e7e05590803a656e46183bc94642ab8R209), we initialized like the following. 
```scala
val requestedDataColIds = requestedColIds ++ Array.fill(partitionSchema.length)(-1)
val requestedPartitionColIds = Array.fill(requiredSchema.length)(-1) ++ Range(0, partitionSchema.length)
```
So, logically, in this `for` loop,  the range of `i` satisfying `requestedPartitionColIds != -1` seems to be filled with `Array.fill(partitionSchema.length)(-1)`? Did I understand correct?

Reply:
Yes. This is because in https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L191
the require schema always filter out the all partition columns.

Now It can be easily fixed in ORC V2, but to fix `FileFormat` it may affect the Parquet reader as well. 
In this PR, I will check the `requestedPartitionColIds` in `requiredSchema`, so that it will be easier if the someday the improvement is made for `FileFormat`.

Reply:
Thanks.

Reply:
The suggested test suite also covers this logic.